proposals/0052-svg-components.md

```diff
@@ -0,0 +1,293 @@
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg'
+---
+
+<Logo />
+
+<!-- Pass width and height to override default size -->
+<Logo width={100} height={100} fill="blue" />
+```
+
+This would generate the following output:
+
+```astro
+<svg width="24" height="24" role="img">
+    <!-- SVG Content -->
+</svg> 
+<svg width="100" height="100" role="img" fill="blue">
+    <!-- SVG Content -->
+</svg> 
+```
+
+### Sizing
+
+To simplify the process of setting the dimensions of SVG components, this proposal introduces a `size` prop that developers can use to uniformly scale the width and height of an SVG. This prop provides a convenient shorthand for situations where both width and height need to be set to the same value, offering a more concise API.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg';
+---
+
+<Logo />
+
+<!-- Using the size prop to set both width and height -->
+<Logo size={48} />
+```
+
+This would generate the following output:
+
+```astro
+<svg width="24" height="24" role="img">
+    <!-- SVG Content -->
+</svg>
+<svg width="48" height="48" role="img">
+    <!-- SVG Content -->
+</svg> 
+```
+
+Caution: This property will not apply if you have set a `height` and/or `width` prop.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg';
+---
+<Logo />
+<Logo size={50} width={20} />
+<Logo width={20} size={50} />
+```
+
+This would generate the following output:
+
+```astro
+<svg width="24" height="24" role="img">
+    <!-- SVG Content -->
+</svg>
+<svg width="20" height="24" role="img">
+    <!-- SVG Content -->
+</svg>
+<svg width="20" height="24" role="img">
+    <!-- SVG Content -->
+</svg> 
+```
+
+## Accessibility Considerations
+
+One key aspect of this design is ensuring that the SVG components generated from `.svg` files are accessible by default, following best practices for web accessibility. This includes the following considerations:
+
+### ARIA Attributes
+
+- **Role Attribute:** By default, Astro could set `role="img"` on SVGs when they are used for non-decorative purposes.
+- **Title Element:** If the `.svg` file needs and accessible title/label, Astro can inject a `<title>` element.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg'
+---
+
+<!-- Pass an accessible title that will be injected  -->
+<Logo title="Company Logo" />
+```
+
+This would generate the following output:
+
+
+```astro
+<svg width="24" height="24" role="img">
+    <title>Company Logo</title>
+    <!-- SVG Content -->
+</svg> 
+```
+
+While this proposal offers strong defaults for accessibility, it is essential to give developers full control. Any automatically generated accessibility attributes (aria-label, role, etc.) should be overridable by the developer at the component level, ensuring that specific use cases or custom behaviors can be supported.
```

Comment:
This is a good point. Our goal was to provide strong defaults for a11y. This (`role="img"`) and `<title>`  come over from `astro-icon`. From everything I have read, `role="img"` and `<title>` is a solid default for svgs. 

Not sure how we could make the DX better here. @natemoo-re might have some insight on this as well.

**references:**
[mdn](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/img_role#svg_and_roleimg)
[accessible SVGs](https://www.smashingmagazine.com/2021/05/accessible-svg-patterns-comparison/#best-in-show-1)
[dequeue](https://www.deque.com/blog/creating-accessible-svgs/)
[w3c](https://www.w3.org/TR/2000/NOTE-SVG-access-20000807/#Equivalent)
[css tricks](https://css-tricks.com/accessible-svgs/)

Reply:
> From everything I have read, `role="img"` and `<title>` is a solid default for svgs.

But it's not the default currently. By default, only `role="img"` is added. The `title` is optional.

And while `role="img"`+`<title>` is fine for labelling a static SVG, my comment is about the implementation in Astro.

1. If the SVG contains meaningful content inside it, `role="img"` may hide it by default. This is surprising behavior.
2. The `title` attribute gets confusingly converted into a `<title>` element. This is surprising behavior.
   - The `title` attribute is [widely discouraged](https://www.24a11y.com/2017/the-trials-and-tribulations-of-the-title-attribute/), so developers may feel uneasy using it and leave the SVG unlabelled.

It may have been fine for a third-party library like `astro-icon`, but a framework should do the least surprising thing. So I would suggest making `role="img"` opt-in, and allow `<title>` to be passed as children.

```astro
---
import MySvg from "somewhere.svg";
---
<MySvg role="img">
  <title>The label</title>
</MySvg>
```

This is more predictable, more flexible, and also addresses my point above where both `role="img"` and `aria-hidden="true"` were present.

(Side note: in my personal experience, it's far more common to hide the SVG using `aria-hidden="true"` and instead make sure to label the encompassing element. The label is often dependent on the surrounding context, and the tooltip added by `<title>` is usually undesirable as well.)

Reply:
Agree with @mayank99 here. For various reasons I'll almost always have markup that looks something like the following (abbreviated for clarity):

```html
<span>
  Text label
  <svg aria-hidden="true">...</svg>
</span>
```

(Substitute button, a, etc. for span as appropriate, and potentially wrap the label in `visually-hidden` class if necessary.)

In fact one of the most common accessibility remediations I've had to apply to sites using `astro-icon` is going through adding `aria-hidden="true"` to everywhere the icon is used.

Reply:
I do think I agree with you. Would still like to hear @natemoo-re and @matthewp 's perspectives as well.

Reply:
Just realized this ties back to my unresolved question: 
https://github.com/withastro/roadmap/pull/1035/files#diff-ab3122e95078e177eab4f364380c44e582e73c496b98809d7d5e966f9d09e1f1R325-R326

Reply:
I'm not sure about `role="img"`, I've never used that on an SVG so I can't even say what it does. So no opinion from me.

Moving `title="title"` to `<title>title</title>` does make sense for me. That's how you give hover context to SVGs. It's not required in the component is it? If not and you want to label your icons in a different way, can't you just not use it?

Reply:
If the `title` prop doesn't set the `<title>` then where are you expecting it to go? On the SVG element itself? Is there a use-case for that?

Reply:
> If the `title` prop doesn't set the `<title>` then where are you expecting it to go? On the SVG element itself? Is there a use-case for that?

That’s just the default HTML `title` attribute. Arguably there are no good use cases for using the `title` attribute on any elements, but I think the argument here is it is strange to special case a valid HTML attribute and give it unexpected behaviour. If I were using an imported SVG and added `title="Something"`, it feels unlikely I’d do that and guess “I bet this adds a `<title>` element.”

_Plus_ using `<title>` is not really a best practice for inline SVGs vs using other accessible labelling approaches, so it feels odd to add special privileged support for it.

Reply:
We shouldn't let [perfect be the enemy of good](https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good). Whether there is some other technique that's more accessible or not, setting `<title>` does allow AT to announcement the SVG, it does allow mouse hovers to show the title. There are better than not setting a title at all. 

Now if there's some reason why people want to set `title` on an SVG and we are now preventing that, then maybe that's a reason not to do this. But I haven't heard that yet.

Reply:
> There are better than not setting a title at all.

It's very hard to determine this without context. We use `astro-icon` on most of our sites and I haven't seen a case where `<title>` is a good idea. The fact this feature existed also made no difference to those sites shipping with missing accessible labels for icons.

I think the burden of evidence lies with proving why special handling is justified. Imagine we proposed the component should have a `border` attribute that would inject a `<rect>` the size of the SVG. We'd have to show why this nonstandard behaviour was beneficial enough to justify subverting how HTML works and better than other techniques for applying a border.

In this case:
- use of `<title>` is not considered best practice (there are much better ways to support AT and its mouse hover behaviour is inaccessible to everyone without a pointer device)
- users are already able to add `<title>` directly to SVG files if needed
- we undermine people learning how standard HTML works by making it _appear_ like the `title` attribute works on SVGs the same way as for HTML elements (they'd have to inspect source to notice the underlying markup changed)

To me it's not about perfect vs good, it's about better. And `<title>` is worse than other options (`aria-label`, a visible `<figcaption>`, visibly hidden accompanying text would all be better). Everyone is still free to use it in their SVG files, but why make an API that privileges an approach that has existing better alternatives?

Reply:
What I'm missing here from the conversation are specifics. Is there an article that explicitly lays out the different approaches and the specific strengths and weaknesses? There was an article shared about about the title attribute, but we're talking about the `<title>` element on an SVG.

I don't doubt that there are better approaches but without seeing them we can't discuss the tradeoffs. Also, even if some other approaches are better that doesn't mean we shouldn't provide a convenient way to set the `<title>` unless it's actually harmful, but I haven't seen that be suggested yet. It's not Astro's role to say that you must only do things the best possible way. Adding convenient methods to do something that's good seems like something we should be doing.

But again, without seeing the alternates and knowing the specific tradeoffs it's hard to know if it's worth the addition or not.

Reply:
> without seeing the alternates and knowing the specific tradeoffs it's hard to know if it's worth the addition or not.

The following approaches are all better. I mentioned them in the previous comment, but I’ll go into more detail about why here.

1. Using a `<figure>` and `<figcaption>` around your SVG.

   **When to use?**
   For illustrations, and other contentful graphics that require a caption.

   **Why is it better?**
   - It accessibly labels the SVG in a way that is visible for sighted users and semantically accessible via the accessibility tree to all users of assistive technology.
   - It is also a regular text node, so tools like browser translation software interact with it as expected.

2. Using accompanying text inside a wrapper element (potentially visually hidden).

   **When to use?**
   Usually this pattern is used in links with complementary icons, icon buttons, etc. The icon itself should be `aria-hidden="true"` in this case.

   **Why is this better?**
   - Like with `<figcaption>` the regular text node behaves as expected with browser translations.
   - It also has the same the benefit of flexibility around whether the text is visually hidden or not.
   - This pattern reduces verbosity. When announcing an SVG with `<title>My cool drawing</title>`, most screenreaders announce something like “My cool drawing, image”. In UI contexts this can be disruptive. Imagine e.g. icon links announced as “link, GitHub, image”, “link, Discord, image”, “link, Bluesky, image”. The image role is completely superfluous here, and worsens UX for AT users.

3. Using an `aria-label` attribute

   **When to use?**
   When neither of the above are feasible.

   **Why is this better?**
   - Because it is explicit about what it does. I’m 100% sure there are subtleties around differences between `title` and `aria-label`, but for the most part they will behave the same. But `aria-label` (and its counterpart `aria-labelledby`) is explicitly designed for applying an accessible label without any of `title`’s weirdness. In practice, I’d avoid both when possible.

>  It's not Astro's role to say that you must only do things the best possible way. 

I agree with this, but as I mentioned above:

> This will also impact documentation. Because this behaviour is nonstandard, we will have to document it. That means our docs would prominently feature a way of accessibly labelling SVGs that is not the best way.

I’m not saying no-one’s allowed to use `<title>`. I’m saying `<title>` is not a good enough method to enable using it in a nonstandard way, which is a way of _promoting_ this usage. Want to add it to your SVG files? Absolutely. Go ahead!

> `aria-label` doesn't show the hover title. So it seems that `<title>` is a better choice there.

I didn’t include this in my criteria above, because globally >50% of all web traffic is completely unable to trigger a title tooltip on hover. Depending on that behaviour is a bad practice that excludes users of touch screens as well as others like keyboard users. Of course, it’s a free world, so people can still use `<title>` if they desperately want a poorly supported tooltip for their personal side project, but it’s not a behaviour that is valuable enough to promote its usage.

Reply:
Thanks for the breakdown of alternatives! I think we're getting to the heart of the issue now, and that's about use-cases and design patterns. I have thoughts to drill down further, but first I want to hit on this point which I think is unrelated:

> I’m not saying no-one’s allowed to use <title>. I’m saying <title> is not a good enough method to enable using it in a nonstandard way, which is a way of promoting this usage. Want to add it to your SVG files? Absolutely. Go ahead!

I don't buy the non-standard argument some have made. Of course it's non-standard, but so is importing an SVG, so are Titlecase components, so is `.astro`. syntax. This is not a standards body, we're defining our own API.

I don't think there's an expectation of props mapping directly to `<svg>` element props. We have lots of components in Astro core that have their own props APIs. The `<Code />` component doesn't map directly to a `<code>` element, nor does `<Image />` map to an `<img />`. Generally speaking, components define their API in frameworks.

That this particular component spreads unknown props onto the `<svg>` element is a positive thing that makes it more powerful; but it shouldn't be a constraint we place on ourselves not to make the API better where appropriate.

Reply:
> it shouldn't be a constraint we place on ourselves not to make the API better where appropriate

I agree with this but with the emphasis on "where appropriate". We aim to design features so Astro users fall into the "pit of success". So it is appropriate to extend and enhance default attributes to make best practices easy.

The RFC includes an ["Accessibility considerations"](https://github.com/stramel/roadmap/blob/built-in-svg-component-stage-3/proposals/0052-svg-components.md#accessibility-considerations) section, which I applaud, just like this whole proposal and implementation. This is a great feature and @stramel has done valuable work pushing it forwards.

But the accessibility section does not provide evidence for why the two choices are best practices. Myself and others have provided significant evidence as to why they may not in fact be best practices. I'd be happy to try and work on alternative API proposals if that's desirable, because I acknowledge there isn't a simple answer, but I think they could also follow later similar to how we have tackled responsive image attributes recently.

Reply:
Thanks for all the great conversation! Really enjoying reading through everyone's perspective!

I'm not tied to the current implementation. If most people are not using `role="img"` and `title` prop, I would be open to switching to a more minimal approach. I do think that allowing children would also result in a good developer experience and potentially better a11y. 

I like @mayank99's proposal for a more minimum approach. 

```astro
---
import MySvg from "somewhere.svg";
---
<MySvg role="img">
  <title>The label</title>
</MySvg>
```

would result in

```svg
<svg role="img">
	<title>The label</title>
	<path d="" />
</svg>
```

This approach does move the a11y to the developer but provides more control.

Reply:
Hey @stramel I think we should continue this conversation before making any changes. I haven't had an opportunity to respond to the meat of @delucis's latest comment but I plan on doing so soon.

Reply:
@delucis I don't find appeals to authority or calling something a "best practice" to be a persuasive argument. It's good to know what experts think on a subject but we have to weigh many different things here, not just the best possible a11y solution (which we can't provide). Similarly sometimes security experts feel strongly about something (see [here](https://github.com/withastro/astro/pull/11535#issuecomment-2248111384) as an example), but that doesn't mean it's the only consideration; it's one of many. So happy to keep discussing the trade-offs and drill down further into the problem.

----

To drill down further, your list of approaches helps clarify what use-case we're really talking about here. You wouldn't use the title when figure/figcaption is an option, so not for illustrations. We're really just talking about icons, and specifically the types of icons that do not have accompanying test.

Here are some icons from Bluesky that illustrate the use-case:

<img width="167" alt="Icons from Bluesky to add images/videos to a message" src="https://github.com/user-attachments/assets/4892c533-695b-4112-aa65-7de24bd6da88" />

The implementation here is using `aria-label`. As such it's announced correctly. But there is no tooltip available, so unless you already know what these icons mean, you don't get any sort of hinted if you are a sighted user.

This seems clearly worse than the alternative of `<title>` to me and I guess that's where we're not seeing eye-to-eye. I understand your point about the tooltip not working on mobile or for keyboard users. But I don't understand why you feel it's preferable to exclude sighted mouse users.

The best solution would be to use a tooltip library and make it work for keyboard, mobile, and mouse users. But that's not an easy solution. That requires a great deal of work; work that a company of Bluesky's pedigree decided to put off. To me that's a pretty clear indication that `<title>` is a small improvement, and so making it easy to make a small improvement still feels like a good thing to provide.

Reply:
@matthewp I'm ok with dismissing the earlier points about nonstandard handling of the `title` prop if staying close to HTML is not a goal for Astro. However, I would still like to reiterate one other point which wasn't addressed, since we are laser-focused on discussing tooltips for mouse users.

Astro will set `role="img"` on the imported SVG component. This makes the SVG be exposed as an _unlabelled_ image in the accessibility tree (to assistive tech, like screen-readers). This is worse than the default, where the SVG contents would be exposed unchanged (which means decorative icons get skipped by default, and any meaningful content inside the SVG gets exposed as-is).

---

> You wouldn't use the title when figure/figcaption is an option, so not for illustrations.

If we consider the illustration example, `role="img"` worsens the default experience provided by figure/figcaption. Screen-reader users can access the caption, but when they drill down into the figure, they will find an unlabelled image (which is announced as just "image"). And this image role is also _hiding_ potentially useful details inside the SVG.

---

> The best solution would be to use a tooltip library and make it work for keyboard, mobile, and mouse users.

When the consumer brings in a library to add tooltips to the icon-buttons, these tooltips would potentially conflict with the tooltips added by Astro's `<title>`. At that point, the consumer might remove the `title` prop, leaving the SVG unlabelled in the accessibility tree (a generic "image").

The best way to label an icon-button is using [visually-hidden text](https://www.scottohara.me/blog/2017/04/14/inclusively-hidden.html) inside the button. This is good because the visually-hidden can be automatically translated ([unlike `aria-label`](https://adrianroselli.com/2019/11/aria-label-does-not-translate.html)).

So the resulting markup might look something like this:

```jsx
<button data-library-tooltip="Upload a file">
  <ImportedSvg />
  <span class="visually-hidden">Upload a file</span>
</button>
```

Because Astro is adding a `role="img"`, the button will be exposed to assistive-technology users as "Button, image upload a file". The "image" should _not_ be part of the button label, and it creates a potential barrier for voice-control users.

This issue can be fixed by adding `aria-hidden="true"` to the imported SVG. But the issue wouldn't even occur if Astro didn't add `role="img"` in the first place.

> so making it easy to make a small improvement still feels like a good thing to provide.

I'm trying to understand what you mean by "making it easy". It's already easy enough if Astro doesn't do anything special and instead just lets consumers pass attributes and children.

These are both equally easy, except one requires learning a new framework-specific API.

```jsx
<MySvg title="The label" />
```

```jsx
<MySvg>
  <title>The label</title>
</MySvg>
```

The goal should be to **make it easy to do the right thing**. Acknowledging that `<title>` may be useful for some cases, Astro is making it more difficult to do the right thing in many other cases, without enough justification.

Reply:
@mayank99 I'm +1 on removing the `role="img"`. The only area of disagreement here is the title sugar.

AFAIK the options we are debating here are:

- `<title>`
- Visually hidden text + tooltip library.

I can agree that the latter is better, but what I don't think is being acknowledged here is that it is significantly more work. Supporting `title` prop doesn't discourage you to use a better option. It just makes it slightly easier to do something that's better than an `aria-label` by itself (which is what the Bluesky example is doing).

Reply:
The `<title>` is meant to be used together with `role="img"`. Without the `role="img"`, `<title>` does not act as a label for the SVG, but rather exists as its own text element.

If you look at the various links shared [above](https://github.com/withastro/roadmap/pull/1035#discussion_r1868795150), they all make use of `role="img"` along with the `<title>`, and additionally recommend using `aria-labelledby` as well (I think it used to not work at all without `aria-labelledby`, but support may have improved since then). 

So the expected markup for an accessible SVG looks like this (to be announced as "image, title":

```jsx
<svg role="img" aria-labelledby={uniqueId}>
  <title id={uniqueId}>{title}</title>
  <path d="…" />
</svg>
```

If Astro is doing special handling of the `title` prop, I expect it to produce the above (at minimum, the `role="img"` needs to be present, because SVGs do not have an implicit role).

Reply:
I've tested this example in both OSX VoiceOver and Windows Narrator and they announce exactly the same (no `role="img"` used):

```html
<article>
  <h1>&lt;title&gt;</h1>
  <button type="button">
  <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-anchor">
    <title>Hello title</title>
  <path d="M12 22V8"/>
  <path d="M5 12H2a10 10 0 0 0 20 0h-3"/>
  <circle cx="12" cy="5" r="3"/>
</svg>
  </button>
</article>

<article>
  <h1>aria-label</h1>
  <button type="button">
  <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-anchor" aria-label="Hello aria">
  <path d="M12 22V8"/>
  <path d="M5 12H2a10 10 0 0 0 20 0h-3"/>
  <circle cx="12" cy="5" r="3"/>
</svg>
  </button>
</article>
```

Reply:
Testing on a handful of screen-reader+browser combinations is [not enough](https://adrianroselli.com/2022/11/your-accessibility-claims-are-wrong-unless.html) to determine if a pattern will work everywhere. There is no guarantee that if it works with one set of technologies, that it will also work everywhere else.

If you look at the [article linked above](https://www.smashingmagazine.com/2021/05/accessible-svg-patterns-comparison), `role="img"` + `<title>` (pattern 5) was [tested](https://codepen.io/smashingmag/pen/YzZQBwG) to work across:

- macOS VoiceOver + Safari/Chrome/Firefox/Edge
- iOS VoiceOver + Safari/Chrome
- Windows NVDA + Chrome/Firefox/Edge
- Android Talkback + Chrome/Firefox

I wouldn't diverge from this thoroughly-tested and repeatedly-recommended pattern.

Reply:
I'll test with all of those combination and report back.

I'm not tied to any particular prop name, no. I'm not even tied to this behavior, I'm only adamant that we understand the trade-offs ourselves.

Reply:
I tested with these combinations:

- OSX VoiceOver + Safari/Chrome/Firefox/Edge
- Windows Narrator + Chrome/Firefox/Edge
- Windows NVDA + Chrome/Firefox/Edge
- Android Talkback + Chrome/Firefox

They work the same in all of those combinations *except* OSX + Safari. Safari doesn't announce the `<title>` version. It also announces `aria-label` a little weirdly, announcing it twice: "Hello aria, button, Hello aria, article".

What if we made the prop set both `aria-label` and `<title>`? That way you get voice over on all browsers, and you get the tooltip on all browsers. This would also be a good reason to change the prop name to something like `label` as @mayank99 suggested.

Reply:
> What if we made the prop set both `aria-label` and `<title>`?

I would advise against that for multiple reasons:
1. [`aria-label` is a code smell](https://ericwbailey.website/published/aria-label-is-a-code-smell/) and should be avoided in general. Among other issues, it [doesn't translate](https://adrianroselli.com/2019/11/aria-label-does-not-translate.html). A better alternative would be `aria-labelledby`, but even that would not be advisable without `role="img"` (see below).
2. The [SVG AAM spec](https://www.w3.org/TR/svg-aam-1.0/#details-id-66) lists `graphics-document` as the default role for `<svg>`, so `aria-label`/`aria-labelledby` is technically allowed; however it's unclear how `graphics-document` should be exposed, so it will have inconsistencies and poor support (e.g. see [nvda#13863](https://redirect.github.com/nvaccess/nvda/issues/13863) for example). Without a reliable default role, the SVG may also be treated like a static element, and [`aria-label` should _not_ be used on static elements](https://benmyers.dev/blog/dont-use-aria-label-on-static-text-elements/).

Despite testing across a few different browser/SR combinations, it still only covers a very small surface area of the ways a user might interact with the SVG. (Remember also that screen-readers are only one form of assistive technology).

Your test case also specifically only focused on SVGs inside a `<button>`, so it's not even testing the accessibility of the SVG itself, but rather that of the button. This distinction is important because the `button` role [flattens](https://w3c.github.io/aria/#childrenArePresentational) any semantic information inside it into a single string which gets used for labelling.

While I'm happy that `role="img"` is being dropped as a default, I think we might be going too far in the other direction by dropping it for altogether. So I would reiterate [my earlier suggestion](https://github.com/withastro/roadmap/pull/1035#discussion_r1884608848): _conditionally add `role="img"` only if the `label` prop is passed_.

Alternatively, do nothing (i.e. ask users to manually add `role`/`aria-label`/`<title>`/whatever else) like I've been suggesting from the [very beginning](https://github.com/withastro/roadmap/pull/1035#discussion_r1869743984).

---

> I'm only adamant that we understand the trade-offs ourselves.

What are the trade-offs? I mentioned [earlier](https://github.com/withastro/roadmap/pull/1035#discussion_r1884542492) that "it's already easy enough if Astro doesn't do anything special and instead just lets consumers pass attributes and children". It avoids all the potential issues that come with Astro making the wrong decision.

The thing I'm hoping to really get across is that this is quite a nuanced topic, and any decision should be made after lots of careful consideration, as there is potential to do real harm.

Reply:
I know it's a nuanced topic, that's why I'm attempting to have a nuanced conversation. Calling something "code-smell" and linking to hyperbolic articles is not displaying nuance. Let's please do so and stick to observable facts and specifics.

> Your test case also specifically only focused on SVGs inside a <button>, so it's not even testing the accessibility of the SVG itself, but rather that of the button. This distinction is important because the button role [flattens](https://w3c.github.io/aria/#childrenArePresentational) any semantic information inside it into a single string which gets used for labelling.

This is the use-case we're discussing. Inside of `<button>` or inside of a `<a>`, both which have roles and will announce (tested). It's ok for a prop to exist to aid in one use-case without it being as useful to others.

If there are other AT to test this with please let me know what those are and I'll do so.

I'm still not sure what is going on with Safari here and need to dig into it deeper. It's also not announcing with `role="img"` so there might be something else wrong with my page.

Reply:
The following ("pattern 8" mentioned [above](https://github.com/withastro/roadmap/pull/1035#discussion_r1884586782)) should work in VoiceOver+Safari and everywhere else:

```html
<button>
  <svg role="img" aria-labelledby="unique-id">
    <title id="unique-id">Favorite</title>
  </svg>
</button>
```

There are lots of other assistive technologies, like screen magnifiers, refreshable braille displays, switch controls, sip-and-puff devices. You can take a look at [how people with disabilities use the web](https://www.w3.org/WAI/people-use-web/user-stories/) to better understand the assistive technologies they rely on. An easy one to test with (and one that's directly relevant here) is [speech recognition software](https://tetralogical.com/blog/2021/11/15/browsing-with-speech-recognition/).

Reply:
Tested and `aria-labelledby` does make it work. No `role` needed. This seems like a pretty good combination to me, you get:

- Voiceover
- Mouse hover tooltip
- Content in the page so translation works

And by omitting `role="img"` it doesn't announce it to be an image. What are the downsides to doing this?

Reply:
@stramel As no one is happen with the current behavior, if you wanted to remove this from the RFC for now, while we continue to try and reach a consensus on a direction, that would be fine with me.

Reply:
Not sure it's too helpful, but I would not expect an SVG integration to muck with a11y settings as a general-purpose tool. astro-icon makes more sense since it's an icon library, but as a framework integration for svgs as icons and as illustrations, I would expect control to be handed to me as the developer to implement my SVG's correctly. I'm not sure there's a one-size fits all approach that works for all usecases and doesn't step on toes.

Maybe an approach closer to `<Image>` where `alt` is a required prop would be prudent? Could write custom lints that require title or aria-label or aria-hidden to be set, but that might be a bit complicated.